### PR TITLE
move all host functions to wasm

### LINKED ISSUE #1308: move all custom host functions 🫠 to wasm
Screenshot 2022-07-12 at 3 29 25 PM

### PR DESCRIPTION
**Pull request by seunlanlege**
closes #1308

### CODE REVIEW

File: frame/ibc/ibc-primitives/src/runtime_interface.rs [hunk 1]
```diff
@@ -12,68 +12,66 @@ pub enum SS58CodecError {
...
+	// 	.map_err(|_| SS58CodecError::InvalidString)
+}
 
-	fn account_id_to_ss58(bytes: [u8; 32]) -> Result<Vec<u8>, SS58CodecError> {
-		use sp_core::crypto::Ss58Codec;
-		let account_id = AccountId32::new(bytes);
-		Ok(account_id.to_ss58check().as_bytes().to_vec())
-	}
+pub fn account_id_to_ss58(bytes: [u8; 32]) -> Result<Vec<u8>, SS58CodecError> {
+	use sp_core::crypto::Ss58Codec;
```
[the remark this answers is not in the document]
Comment (Wizdave97) [reply frame/ibc/ibc-primitives/src/runtime_interface.rs:74]: We could remove ss58 support and stick to hex strings until we implement it in  wasm.

[the remark this answers is not in the document]
Comment (seunlanlege) [reply frame/ibc/ibc-primitives/src/runtime_interface.rs:74]: hex strings make it harder to understand onchain relationship with address. Will implement ss58 in wasm for this pr

File: frame/ibc/ibc-primitives/src/runtime_interface.rs [hunk 2]
```diff
@@ -164,3 +163,16 @@ fn ss58hash(data: &[u8]) -> Vec<u8> {
 	pre_image.extend(data);
 	sp_io::hashing::blake2_256(&pre_image).to_vec()
 }
+
+#[derive(Debug)]
+pub struct Blake2HostFunctionsHasher;
```
Comment (Wizdave97) [inline frame/ibc/ibc-primitives/src/runtime_interface.rs:168]: We can use the `sp_runtime::traits::BlakeTwo256`, it already implements this.

Comment (seunlanlege) [reply frame/ibc/ibc-primitives/src/runtime_interface.rs:168]: good catch

File: frame/ibc/ibc-primitives/src/runtime_interface.rs [hunk 3]
```diff
@@ -12,68 +15,149 @@ pub enum SS58CodecError {
...
+}
+
+pub fn account_id_to_ss58(bytes: [u8; 32]) -> Result<Vec<u8>, SS58CodecError> {
+	let account_id = AccountId32::new(bytes);
+	let encoded = to_ss58check_with_version(&account_id, 49); // todo: this should be based on the runtime.
+	Ok(encoded.as_bytes().to_vec())
+}
+
+// lifted directly from sp-core
+fn to_ss58check_with_version(account_id: &AccountId32, version: u16) -> String {
```
Comment (blasrodri) [inline frame/ibc/ibc-primitives/src/runtime_interface.rs:80]: shouldn't we have a test for this?

Comment (blasrodri) [reply frame/ibc/ibc-primitives/src/runtime_interface.rs:80]: or because it's a verbatim copy from sp-core, we don't need?

File: node/src/service.rs [hunk 4]
```diff
@@ -39,6 +39,8 @@ use crate::{
 pub struct PicassoExecutor;
 
 impl sc_executor::NativeExecutionDispatch for PicassoExecutor {
+	/// This should not be extended under any circumstances or we risk bricking our chain
```
Comment (blasrodri) [inline node/src/service.rs:42]: can you ELI5 when you say "this should not be extended?"

File: frame/ibc/ibc-primitives/src/runtime_interface.rs [hunk 5]
```diff
@@ -12,68 +15,149 @@ pub enum SS58CodecError {
...
+		_ => return Err(PublicError::InvalidPrefix),
+	};
+	if data.len() != prefix_len + body_len + CHECKSUM_LEN {
+		return Err(PublicError::BadLength)
 	}
+	// not available in std
+	// let format = ident.into();
+	// if !AccountId32::format_is_allowed(format) {
+	// 	return Err(PublicError::FormatNotAllowed)
+	// }
```
Comment (blasrodri) [inline frame/ibc/ibc-primitives/src/runtime_interface.rs:144]:
```suggestion
```

**Review by blasrodri — APPROVED** [review]
Aside from https://github.com/ComposableFi/composable/issues/1308#issuecomment-1191777221 and small nits, LGTM

**Review by Wizdave97 — APPROVED** [review]
Looks Good

File: node/src/service.rs (hunk 4 above)
Comment (seunlanlege) [reply node/src/service.rs:42]: This is where host functions can be extended, by supplying a tuple of them.

File: frame/ibc/ibc-primitives/src/runtime_interface.rs (hunk 3 above)
Comment (seunlanlege) [reply frame/ibc/ibc-primitives/src/runtime_interface.rs:80]: good catch, will copy the tests

File: frame/ibc/src/lib.rs [hunk 6]
```diff
@@ -126,7 +126,9 @@ pub mod pallet {
 	};
 	use frame_system::pallet_prelude::*;
 	use ibc::core::{
-		ics02_client::msgs::create_client::TYPE_URL as CREATE_CLIENT_TYPE_URL,
+		ics02_client::msgs::{
+			create_client, create_client::TYPE_URL as CREATE_CLIENT_TYPE_URL, update_client,
```
Comment (seunlanlege) [inline frame/ibc/src/lib.rs:130]:
```suggestion
			create_client, create_client::TYPE_URL as CREATE_CLIENT_TYPE_URL,
```

File: frame/ibc/src/lib.rs [hunk 7]
```diff
@@ -126,7 +126,9 @@ pub mod pallet {
 	};
 	use frame_system::pallet_prelude::*;
 	use ibc::core::{
-		ics02_client::msgs::create_client::TYPE_URL as CREATE_CLIENT_TYPE_URL,
+		ics02_client::msgs::{
+			create_client, create_client::TYPE_URL as CREATE_CLIENT_TYPE_URL,
```
Comment (seunlanlege) [inline frame/ibc/src/lib.rs:130]:
```suggestion
			create_client::TYPE_URL as CREATE_CLIENT_TYPE_URL,
```
